### PR TITLE
1.0.0-alpha DL4J support

### PR DESCRIPTION
**Pull request by enragedginger**
DL4J 1.0.0-alpha just released. There's been a large number of API changes. I started working on supporting those. I've left my comment inline below. Overall, i think we can pull it off, we just might need to restructure the way our NN data structures are organized. In general, I think we would benefit from moving more towards maps instead of vectors with keys and values interspersed everywhere.

### CODE REVIEW

**Comment by hswick** [comment]
Could you give an example where the current NN config doesn't fit well with the latest release?

In terms of the syntax, I would like to keep it as close to [hiccup](https://github.com/weavejester/hiccup/wiki/Syntax) style as possible. Since that is a standard of sorts (for once in the Clojure ecosystem). I would say currently the config deviates from hiccup in that it doesn't use maps, and maybe doesn't nest as much.  I am open to using maps for the inner layer configs. But I would like to keep the layers and the global config as vectors.

**Comment by enragedginger** [comment]
I spent some time thinking about this. I love hiccup style syntax for XML / HTML as it makes perfect sense for that domain. For other things, I don't think it's a perfect fit. For this particular library, it works and I think it can continue to work. I think my issue with the way it's implemented here is that the parser isn't fully recursive. I've done a couple of things in this PR:
1. I've updated the tests based on some questions I asked over on the [DL4J GitHub](https://github.com/deeplearning4j/deeplearning4j/issues/4903).
2. I've added working implementations of things like `Adam` and `Nesterovs` updaters to the tests as well as an example of what I'd like to see this code look like in the future. I think you're infinitely more familiar with the underlying code of this library that's actually parsing those options, so let me know if that's something we should shoot for / if you're able to do it.

For now though, I'm going to move on to testing this library with the DL4J 1.0.0-alpha release with my side projects to see how it all performs.

Thanks for all your work on this!

-Stephen

**Comment by hswick** [comment]
Awesome! You've been killing it! 

I think that Adam and Nesterovs may merit their own helper functions for now.

Here are some ideas as to what they could look like:
```clojure
(ai/adam [:scheduler :map-schedule :schedule-type :epoch ...&args])
```
Where the key - value list maps to the options. Using a vector not a map because order does matter for the nd4j api.

Or if the options aren't too varied we could get away with:
```clojure
(ai/adam :map-schedule :epoch ...&args)
```

Or:
```clojure
(ai/updater :adam :map-schedule :epoch ...&args)

;;with vector (more consistent)
(ai/updater [:adam :map-schedule :epoch ...&args])

;;with map
(ai/updater {:updater :adam :scheduler :map-schedule :schedule-type :epoch :args ...&args})

;;more explicit vector
(ai/updater [:updater :adam :scheduler :map-schedule :schedule-type :epoch :args ...&args])
```

I kind of like using the `ai/updater` the most because it easily maps mentally to the :updater keyword in the config. And then we can supplement the method with documentation so its easy to know which options are available.

A larger discussion about syntax changes may merit its own separate issue, which I would be happy to discuss further

- Harley

**Comment by enragedginger** [comment]
tl;dr I've improved support for updaters and schedules and I've updated the tests to use them. Basically, I went the map route for the most part. The updaters (and schedules) all vary greatly in terms of expected / allowed combinations of parameters. Most of the jutsu.ai core code just calls the set method on the network builders. However, the updaters don't have standard Java POJO set methods for the expected parameters. This means that you have to have the values up front when you construct the object. This isn't really a problem for us. I thought of using vectors with keys and values interspersed but decided against it because (1) I would have just called `apply hash-map` anyways and (2) the order that a user of `jutsu.ai` passes in those parameters doesn't / shouldn't matter.

Anyhow, take a look and let me know what you think. My side ML projects that are using jutsu.ai and DL4J seem to be working splendidly with these changes and the 1.0.0-alpha so far.

**Comment by hswick** [comment]
Wow that was a lot of different options. I see your point

**Comment by hswick** [comment]
Finally, got around to updating the jutsu.matrix version and was able to get the tests to pass.

2 small things before I'd like to merge this:
- build.boot - needs updated versions (I use boot to test/deploy)
- Change bottom of :import to `[]` brackets instead of `()` to keep it consistent

**Comment by enragedginger** [comment]
Okay. Updates are done. Let me know if I missed anything else.
